### PR TITLE
Implement CrashpadInfo stream

### PR DESCRIPTION
**Pull request by jan-auer**
The CrashpadInfo stream is an extension containing additional annotations for
the minidump and its modules.

### CODE REVIEW

**Comment by jan-auer** [comment]
@luser before I continue with this, what do you think about the approach to eagerly parse all structures in the Stream? It looks like this is what we generally do, but it looks quite wasteful to me. 

I'm going to implement the rest of module parsing if what's there now looks good.

**Comment by jan-auer** [comment]
This is now complete and tested on an Electron minidump. Unfortunately that dump is too large to check in. Maybe I can look into creating a smaller file with annotations.

File: src/minidump.rs [hunk 1]
```diff
@@ -1513,6 +1557,200 @@ impl MinidumpAssertion {
...
+            .or(Err(Error::StreamReadFailure))?
+            .to_owned();
+
+        strings.push(string);
+    }
+
+    Ok(strings)
+}
+
+fn read_simple_string_dictionary(
```
Comment (luser) [inline src/minidump.rs:1592]: Can this function be implemented in terms of the existing [`read_stream_list`](https://github.com/luser/rust-minidump/blob/0759d861d60374379388d587d59b1df645a8b056/src/minidump.rs#L631) function? That one does seem to have a little bit of special-casing for alignment but otherwise it reads a length-prefixed array of structs, which is what this does.

File: src/minidump.rs [hunk 2]
```diff
@@ -1513,6 +1557,200 @@ impl MinidumpAssertion {
...
+        let value = read_string_utf8(&mut (entry.value as usize), all, endian)
+            .or(Err(Error::StreamReadFailure))?;
+
+        dictionary.insert(key.to_owned(), value.to_owned());
+    }
+
+    Ok(dictionary)
+}
+
+fn read_annotation_objects(
```
Comment (luser) [inline src/minidump.rs:1626]: Same question here about `read_stream_list`.

File: src/minidump.rs [hunk 3]
```diff
@@ -1513,6 +1557,200 @@ impl MinidumpAssertion {
...
+
+impl<'a> MinidumpStream<'a> for MinidumpCrashpadInfo {
+    const STREAM_TYPE: MINIDUMP_STREAM_TYPE = MINIDUMP_STREAM_TYPE::CrashpadInfoStream;
+
+    fn read(bytes: &'a [u8], all: &'a [u8], endian: scroll::Endian) -> Result<Self, Error> {
+        let raw: md::MINIDUMP_CRASHPAD_INFO = bytes
+            .pread_with(0, endian)
+            .or(Err(Error::StreamReadFailure))?;
+
+        if raw.version > md::MINIDUMP_CRASHPAD_INFO::VERSION {
```
Comment (luser) [inline src/minidump.rs:1735]: This seems incorrect per [the Crashpad docs](https://crashpad.chromium.org/doxygen/structcrashpad_1_1MinidumpCrashpadInfo.html#a481042efad46164dc43d9c824df9e6cf):
> Upon encountering a value greater than kVersion, a reader should assume that the structure’s layout is compatible with the structure defined as having value kVersion.

We have something similar already in the handling of the versioned [`MINIDUMP_MISC_INFO_`](https://github.com/luser/rust-minidump/blob/0759d861d60374379388d587d59b1df645a8b056/src/minidump.rs#L1192) structs.

File: src/minidump.rs [hunk 4]
```diff
@@ -1513,6 +1557,200 @@ impl MinidumpAssertion {
...
+
+        Ok(Self {
+            raw,
+            simple_annotations,
+            module_list,
+        })
+    }
+}
+
+impl MinidumpCrashpadInfo {}
```
Comment (luser) [inline src/minidump.rs:1752]: This can just be omitted if it doesn't contain anything.

File: minidump-common/src/format.rs [hunk 5]
```diff
@@ -214,8 +214,9 @@ pub enum MINIDUMP_STREAM_TYPE {
     /// See ['DSO_DEBUG_64'](struct.DSO_DEBUG_64.html) and
     /// ['DSO_DEBUG_32'](struct.DSO_DEBUG_32.html).
     LinuxDsoDebug = 0x4767000A,
-    /* Crashpad extension types. 0x4350 = "CP"
-     * See Crashpad's minidump/minidump_extensions.h. */
+    /// Crashpad extension types.
```
Comment (luser) [inline minidump-common/src/format.rs:217]: Can you preserve the `0x4350 = "CP"` in the original comment?

**Review by luser — APPROVED** [review]
I have some suggestions and a few questions but overall this seems fine and I'm happy to merge it.

Reading the whole thing up-front ought to be fine, I think. Presumably nobody will read this until they want the data, at which point we might as well read it all. I refactored `MinidumpMemory` to have a lifetime a while ago because memory regions can be sizable and not allocating those bytes seemed reasonable, but most of the other data in the minidump just isn't big enough to really worry about.

File: minidump-common/src/format.rs [hunk 6]
```diff
@@ -1634,3 +1635,213 @@ pub struct DSO_DEBUG_64 {
...
+pub struct MINIDUMP_STRING {
+    /// The length of the #Buffer field in bytes, not including the `NUL` terminator.
+    ///
+    /// This field is interpreted as a byte count, not a count of Unicode code points.
+    pub length: u32,
+    /// The string, encoded in UTF-8, and terminated with a `NUL` byte.
+    pub buffer: Vec<u8>,
+}
+
+impl<'a> scroll::ctx::TryFromCtx<'a, Endian> for MINIDUMP_STRING {
```
[the remark this answers is not in the document]
Comment (jan-auer) [reply minidump-common/src/format.rs:1650]: Agreed. This is also a common pattern in PDBs, where the length field is even offset from the variable-length buffer. It's not too bad to write a manual impl, though :)

**Comment by luser** [comment]
> This is now complete and tested on an Electron minidump. Unfortunately that dump is too large to check in. Maybe I can look into creating a smaller file with annotations.

Conveniently for you, I thought ahead and [wrote this `SynthMinidump` code](https://github.com/luser/rust-minidump/blob/0759d861d60374379388d587d59b1df645a8b056/src/synth_minidump.rs#L13) to generate arbitrary synthetic minidump files in-memory for writing tests. [There are more tests that use it in `minidump.rs`](https://github.com/luser/rust-minidump/blob/0759d861d60374379388d587d59b1df645a8b056/src/minidump.rs#L1701) as well. You can either just write tests using the `SimpleStream` struct defined there to add the Crashpad info stream, or you can go all-in and write a new struct for it and `impl Stream` for it. There are plenty of existing structs you can crib from if you'd like to do that. I'm not hung up on trying to hit 100% test coverage, but having some basic tests of the usual case + some edge cases (empty lists, etc.) would be great.

File: minidump-common/src/format.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jan-auer) [reply minidump-common/src/format.rs:1650]: Also, this reminds me that we ditched `scroll-derive` in pdb for faster compile times. It turned out that scroll takes pretty long to derive, and in many cases the manual derives are easy enough to maintain.

File: minidump-common/src/format.rs [hunk 7]
```diff
@@ -1634,3 +1635,213 @@ pub struct DSO_DEBUG_64 {
...
+    type Error = scroll::Error;
+
+    fn try_from_ctx(src: &[u8], endian: Endian) -> Result<(Self, usize), Self::Error> {
+        let offset = &mut 0;
+        let length: u32 = src.gread_with(offset, endian)?;
+        let data: &[u8] = src.gread_with(offset, length as usize)?;
```
[the remark this answers is not in the document]
Comment (jan-auer) [reply minidump-common/src/format.rs:1656]: This actually shows two separate issues; I'll fix them momentarily.

1) This is dead code. I implemented this and then noticed that the way we read minidumps currently we have helper functions in `minidump.rs`. I'd appreciate some input on what you would prefer at this point. 

In a follow-up, I would probably refactor the minidump module to deserialize into usable types directly. Right now, there is an almost-Rust implementation of the format, and then a wrapper. See [this](https://github.com/willglynn/pdb/blob/547e825dc08fd4fbb883917d3466b0fb4b3560e9/src/symbol/mod.rs#L343-L378) for an example of such such a pattern.

2) This implementation reads too little and the other implementation doesn't validate the NUL byte. Both need fixing.

File: minidump-common/src/format.rs [hunk 8]
```diff
@@ -1634,3 +1635,213 @@ pub struct DSO_DEBUG_64 {
...
+
+/// A variable-length UTF-8-encoded string carried within a minidump file.
+#[derive(Clone)]
+pub struct MINIDUMP_STRING {
+    /// The length of the #Buffer field in bytes, not including the `NUL` terminator.
+    ///
+    /// This field is interpreted as a byte count, not a count of Unicode code points.
+    pub length: u32,
+    /// The string, encoded in UTF-8, and terminated with a `NUL` byte.
+    pub buffer: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (jan-auer) [reply minidump-common/src/format.rs:1647]: The linked code is in a linked crate, however. When you check `format.rs`, then you can see all strings declared there are actually `Vec<u8>`. `minidump.rs` actually reads into a String directly.

See https://github.com/luser/rust-minidump/pull/119#discussion_r543496829, I think that it would make sense to parse into a usable structure right away. However, I'd propose to do that in a separate changeset.

File: src/minidump.rs (hunk 1 above)
Comment (jan-auer) [reply src/minidump.rs:1592]: It can, but that would mean an intermediate allocation of the vector of raw entries that can be avoided by streaming them through. 

Here ideally we'd make `read_stream_list` an iterator so that we can map the raw values, but I think this could be its own PR. Would you prefer me to refactor in this changeset?

**Comment by jan-auer** [comment]
Thanks, @luser! Updated the implementation, pending questions in some of the comments.

> I thought ahead and wrote this SynthMinidump code to generate arbitrary synthetic minidump files

❤️  I'll have a look at that. Most likely, it would make sense to go all the way and `impl Stream`.

File: minidump-common/src/format.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (luser) [reply minidump-common/src/format.rs]: That's a bummer! Making the compiler do this work and not having to maintain boilerplate implementations with all the potential hazards that entails is one of the big selling points for scroll for me. Is it just the fact that you don't have any other dependencies that pull in `syn`, so you eat that whole cost for it? This would be a useful data point in a discussion about Rust improving the state of proc macros.

File: minidump-common/src/format.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (luser) [reply minidump-common/src/format.rs]: > In a follow-up, I would probably refactor the minidump module to deserialize into usable types directly. Right now, there is an almost-Rust implementation of the format, and then a wrapper

I'm conflicted about this. I agree that having two parallel structs is a bit of a pain, but I really like the fact that the structs in `format.rs` map directly to the data on disk for the most part. I don't think you get that same benefit in `pdb` because the PDB file format is quite a bit different. The one similarity is that both formats have data streams that are composed of primitive data types but also contain references to other data types that you need to read to really make use of them, like `MINIDUMP_MODULE_LIST` containing a list of `MINIDUMP_MODULE` objects but then needing to go read the name and CodeView record for each module separately. The APIs in this crate don't necessarily do a great job of leveraging the distinction, but I know that for certain operations on PDB files it's critical to not parse entire data streams up-front, but do so iteratively or using the hashtables in the file for performance.

I'm open to suggestions, especially since you're actively working on the codebase now and I'm not.

File: src/minidump.rs (hunk 1 above)
Comment (luser) [reply src/minidump.rs:1596]: No need to do the refactoring here. If you have an idea of how to make that work feel free to fix it in a followup.

**Comment by luser** [comment]
I'm sure you're already aware, but the `docs` CI job is failing because rustdoc is pedantic about having bare URLs in docs, which should be easy enough to fix. The `macOS-latest, nightly` job is failing the same way on master, so I'm fine with merging this and ignoring that failure. I filed #121 to track that.

File: minidump-common/src/format.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jan-auer) [reply minidump-common/src/format.rs]: Fully agreed. I didn't perform the benchmarks myself but knowing `syn`, it is likely the culprit. The change was made in https://github.com/willglynn/pdb/pull/70.

File: minidump-common/src/format.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jan-auer) [reply minidump-common/src/format.rs]: You got a point here, which would be especially interesting once we implement a Minidump _writer_. In that case, it's definitely useful to maintain raw structs and then have an easy streamlined mapping to the higher-level types. We can definitely draw a lot of inspiration from `goblin` for this.

I'll follow up about this in a separate issue over the next couple of days to collect thoughts.

[the remark this answers is not in the document]
Comment (luser) [reply minidump-common/src/format.rs]: My former Mozilla colleagues [have started work on a Rust crate for minidump writing](https://phabricator.services.mozilla.com/D98744). It doesn't look like it uses this crate directly currently, but maybe vendors parts of it? I've inquired about the reasoning for that.

**Comment by jan-auer** [comment]
@luser I looked into adding tests using SynthMinidump and test assembler. Quite honestly, I can no longer see the point of it... Initially, I agree that some basic coverage is useful, but made a few observations:

Unfortunately, the way the stream is organized doesn't make it easy to be synthesized. I also tried to read the original stream from a Crashpad-generated minidump, but they built it in a way where the strings lie outside of the stream's data range (although adjacent). This leads to a rather brittle implementation of the synthesizing code.

Next, we already have two implementations for all the crashpad extension types, and synthesizing them requires a third implementation. All we're testing is that three implementations agree with each other, but not that either of them is correct. And given the current test setup, I see no way of creating confidence in that correctness.

It would be something different if synthesizing and reading would be one type, as there we could at least test round-tripping. But in this case, it would just end up being two equally complex implementations trying to agree with each other.

I'll keep playing around with this to see if I can get to a satisfying point, but at this moment I'd even opt for not checking in tests that just convey a false sense of safety --

**Comment by luser** [comment]
That's totally fair. I'd like to have some test coverage for this, but if it's a hassle I won't die on that hill. If we can get or create a small test dump using crashpad we can just check it in next to the other test minidumps in the repo. From some brief research, crashpad appears to have [a `generate_dump` tool](https://chromium.googlesource.com/crashpad/crashpad/+/master/tools/generate_dump.md) as well as [a `run_with_crashpad`](https://chromium.googlesource.com/crashpad/crashpad/+/refs/heads/master/tools/run_with_crashpad.md) tool. The latter has commandline arguments to add annotations, which sounds like it'd let you generate dumps with useful data here.

[The tests in `test_minidump.rs`](https://github.com/luser/rust-minidump/blob/4907b5f198b4d3c9ebf8a731a13b2ee9188c648f/tests/test_minidump.rs#L43) use minidumps from the `testdata` directory, so if you commit a new test dump you can add a test there.

**Comment by luser** [comment]
I built crashpad on my Macbook and used it to generate a minidump from a tiny C program that just calls `abort()`. I looked at the result using `minidump_dump` and it has annotations, [so I committed it to the `testdata` directory for you](https://github.com/luser/rust-minidump/blob/master/testdata/simple-crashpad.dmp).

**Comment by jan-auer** [comment]
Thanks, that's excellent. Integration test added to `test_minidump`. Also added a few tests covering more module annotations using the synth module (I still think this is completely over-engineered, but it was a fun exercise). 

Apart from new types, changes in `synth_minidump` mostly revolve around carrying around out-of-band data. Since the only place where we need this is in `List<T>`, there are the following changes:
- `List` no longer requires `DumpSection`, but rather `ListItem: DumpSection`.
- `DumpSection` no longer requires `Into<Section>`. 
- Methods previously requiring `DumpSection` now require `DumpSection + Into<Section>`. 

These changes allow us to implement a new method `ListItem::into_sections() -> (Section, Option<Section>)`, where the first section is compact in-band data and the second section out-of-band data. Each type implementing `ListItem` can again compose of other list items and propagate all their out-of-band data upwards. `List` ultimately concatenates the two together. In our case, this happens in several nesting levels: For CrashpadInfo, ModuleCrashpadInfo and again for each of the annotation lists.

**Comment by jan-auer** [comment]
Merging this now since I believe all review feedback has been addressed. I'll be following up in separate PRs.
